### PR TITLE
Batch Normalization Support

### PR DESCRIPTION
**Pull request by LazaroHurtado**
Full implementation of Batch Normalization for a generic dimension. Pytorch supports BatchNorm for [1D](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html#torch.nn.BatchNorm1d), [2D](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html#torch.nn.BatchNorm2d), and [3D](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm3d.html#torch.nn.BatchNorm3d) mini-batched channeled inputs, this is represented in `tests/batch_normalization.rs`.

The purpose of Batch Normalization is to reduce "[internal covariate shift](https://arxiv.org/pdf/1502.03167.pdf)", in other words, it attempts to normalize each layers input distribution based on two learnable parameters, gamma and beta. [This](https://towardsdatascience.com/batch-norm-explained-visually-how-it-works-and-why-neural-networks-need-it-b18919692739) is a good resource, aside from the actual paper linked, to learn more about BN

### CODE REVIEW

File: src/layers/batch_normalization.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        batch_stat: &Tensor<D>,
+    ) {
+        let scaled_batch_stat = (batch_stat * self.momentum)
+            .into_shape(batch_stat.len())
+            .unwrap();
+
+        let running_stat = celled_running_stat.take();
+        let scaled_running_stat = &running_stat * (1. - self.momentum);
+
+        let new_running_mean = scaled_batch_stat + scaled_running_stat;
```
Comment (LazaroHurtado) [inline src/layers/batch_normalization.rs:58]: Name change required

File: src/layers/batch_normalization.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    where
+        D::Smaller: Dimension<Larger = D>,
+    {
+        let shape = input.shape();
+        let biased = axes.iter().map(|axis| shape[*axis] as f64).product::<f64>();
+
+        let mean = self.iterative_mean(input, axes);
+        let normalized = (input - mean).mapv(|v| v.powf(2.0));
+
+        let var = self.iterative_mean(&normalized, axes);
```
Comment (LazaroHurtado) [inline src/layers/batch_normalization.rs:97]: Using `ndarray`’s `var_axis` would lead to NaN values if unbiased, and 0 if biased, both of which would be wrong. Pytorch allows taking the variance across multiple dimensions, hence letting `n` be the amount of elements in each dimension, which `ndarray` does not support. Therefore, I am using the mean and then handling the biased or unbiased factor in the `match` statement below

File: src/layers/batch_normalization.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        match corrected {
+            true => (var * biased) / (biased - 1.0),
+            false => var,
+        }
+    }
+
+    fn broadcast<D: Dimension>(&self, parameter: &Tensor<Ix1>, ndim: usize) -> Tensor<D> {
```
Comment (LazaroHurtado) [inline src/layers/batch_normalization.rs:104]: Should change to `reshape`

File: src/layers/batch_normalization.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        parameters.append(&mut self.bias.clone().into_raw_vec());
+
+        Tensor::from_vec(parameters)
+    }
+
+    fn forward(&self, input: &Tensor<D>) -> Tensor<D> {
+        let n = input.ndim();
+
+        let mut single_channel_mean_dims = (2..n).collect::<Vec<usize>>();
+        let mut mean_dims = vec![0];
```
Comment (LazaroHurtado) [inline src/layers/batch_normalization.rs:136]: Should change since it is not only used for the mean

**Review by jrego43 — APPROVED** [review]
Looks good to me!

**Review by Demba3 — APPROVED** [review]
